### PR TITLE
feat: lazy loaded esm sources

### PR DESCRIPTION
**Pull request by crowlKats**
This commit adds a new lazy to lazy load internal runtime code
authored in ES modules using "op_lazy_load_esm" op.

The code needs to embedded in the binary to be eliglble for lazy-loading,
and that can be achieved using "lazy_loaded_esm" option on the
"extension!" macro.

Prerequisite for https://github.com/denoland/deno/pull/20812

### CODE REVIEW

File: core/ops_builtin_v8.rs [hunk 1]
```diff
@@ -34,6 +34,48 @@ pub fn op_unref_op(scope: &mut v8::HandleScope, promise_id: i32) {
   context_state.borrow_mut().unrefed_ops.insert(promise_id);
 }
 
+const SPECIFIER: &str = "foobar";
+const SOURCE_CODE: &str = "Deno.core.print('foo\\n')";
+
+#[op2(core)]
+#[global]
+pub fn op_lazy_load_esm(scope: &mut v8::HandleScope) -> Result<v8::Global<v8::Value>, anyhow::Error> {
```
[the remark this answers is not in the document]
Comment (bartlomieju) [reply core/ops_builtin_v8.rs:42]: I would opt to have a separate function called `lazy_load_es_module_included_in_binary` that would accept 2 arguments (`scope`, `specifier`) and let's do everything there - preferably in `core/modules/mod.rs`. Let's keep the op really simple.

File: core/ops_builtin_v8.rs [hunk 2]
```diff
@@ -34,6 +34,48 @@ pub fn op_unref_op(scope: &mut v8::HandleScope, promise_id: i32) {
...
+  let module_map_rc = JsRealm::module_map_from(scope);
+
+  let mod_ns = futures::executor::block_on(async {
+    // false for side module (not main module)
+    let mod_id = module_map_rc
+      .borrow_mut()
+      .new_es_module(scope, false, SPECIFIER.to_string().into(), SOURCE_CODE.to_string().into(), false)
```
[the remark this answers is not in the document]
Comment (littledivy) [reply core/ops_builtin_v8.rs:49]: would it be possible to load from a seperate snapshot of an extension?

File: core/00_primordials.js [hunk 3]
```diff
@@ -613,4 +613,6 @@
   // Provide bootstrap namespace
   globalThis.__bootstrap ??= {};
   globalThis.__bootstrap.primordials = primordials;
+  globalThis.__bootstrap.internals ??= {};
+  globalThis.__bootstrap.internals.primordials = primordials;
```
Comment (bartlomieju) [inline core/00_primordials.js:617]: Ideally we shouldn't need this, figure out a way how we can still access primordials.

File: core/01_core.js [hunk 4]
```diff
@@ -824,8 +824,7 @@
     currentUserCallSite,
   });
 
-  const internals = {};
-  ObjectAssign(globalThis.__bootstrap, { core, internals });
+  ObjectAssign(globalThis.__bootstrap, { core,  });
```
Comment (bartlomieju) [inline core/01_core.js:827]: Unrelated change, revert

File: core/modules/loaders.rs [hunk 5]
```diff
@@ -171,6 +172,59 @@ impl ModuleLoader for ExtModuleLoader {
   }
 }
 
+pub(crate) struct LazyEsmModuleLoader {
```
Comment (bartlomieju) [inline core/modules/loaders.rs:175]: Add a comment on how that works

File: core/modules/loaders.rs [hunk 6]
```diff
@@ -171,6 +172,59 @@ impl ModuleLoader for ExtModuleLoader {
...
+  fn load(
+    &self,
+    specifier: &ModuleSpecifier,
+    _maybe_referrer: Option<&ModuleSpecifier>,
+    _is_dyn_import: bool,
+  ) -> Pin<Box<ModuleSourceFuture>> {
+    let sources = self.sources.borrow();
+    let source = match sources.get(specifier.as_str()) {
+      Some(source) => source,
+      None => return futures::future::err(anyhow!("Specifier \"{}\" was not passed as an extension module and was not included in the snapshot.", specifier)).boxed_local(),
```
Comment (bartlomieju) [inline core/modules/loaders.rs:206]: Error message needs to be updated.

File: core/modules/map.rs [hunk 7]
```diff
@@ -93,7 +93,7 @@ pub(crate) struct ModuleMap {
   pending_dyn_mod_evaluations: RefCell<Vec<DynImportModEvaluate>>,
   pending_dyn_mod_evaluations_pending: Cell<bool>,
   pub(crate) pending_mod_evaluate: RefCell<Option<ModEvaluate>>,
-  data: RefCell<ModuleMapData>,
+  pub(crate) data: RefCell<ModuleMapData>,
```
Comment (bartlomieju) [inline core/modules/map.rs:96]: Try to not expose it.

File: core/modules/map.rs [hunk 8]
```diff
@@ -1240,7 +1240,9 @@ impl ModuleMap {
       .iter()
       .map(|mod_name| (self.get_handle_by_name(mod_name).unwrap(), mod_name))
       .collect::<Vec<_>>();
+    let lazy_esm = self.data.borrow().lazy_esm.clone();
     *self.data.borrow_mut() = ModuleMapData::default();
+    self.data.borrow_mut().lazy_esm = lazy_esm;
```
Comment (bartlomieju) [inline core/modules/map.rs:1245]: This should be addressed by #363

File: core/modules/module_map_data.rs [hunk 9]
```diff
@@ -1,3 +1,4 @@
+use std::cell::RefCell;
```
Comment (bartlomieju) [inline core/modules/module_map_data.rs:1]: Move below the copyright

File: core/modules/module_map_data.rs [hunk 10]
```diff
@@ -128,6 +131,7 @@ pub(crate) struct ModuleMapData {
   /// to evaluate a "synthetic module".
   pub(crate) synthetic_module_value_store:
     HashMap<v8::Global<v8::Module>, v8::Global<v8::Value>>,
+  pub(crate) lazy_esm: Rc<RefCell<HashMap<&'static str, ExtensionFileSource>>>,
```
Comment (bartlomieju) [inline core/modules/module_map_data.rs:134]: Rename to `lazy_esm_sources`

File: core/modules/testdata/lazy_loaded.js [hunk 11]
```diff
@@ -0,0 +1,6 @@
+export const foo = "foo";
```
Comment (bartlomieju) [inline core/modules/testdata/lazy_loaded.js:1]: Format this file before landing

File: core/modules/tests.rs [hunk 12]
```diff
@@ -408,6 +408,28 @@ fn test_mods() {
   assert_eq!(DISPATCH_COUNT.load(Ordering::Relaxed), 1);
 }
 
+#[test]
+fn test_mods_() {
```
Comment (bartlomieju) [inline core/modules/tests.rs:412]: Add a proper name to the test.

File: core/ops_builtin_v8.rs [hunk 13]
```diff
@@ -35,6 +35,69 @@ pub fn op_unref_op(scope: &mut v8::HandleScope, promise_id: i32) {
...
+  scope: &mut v8::HandleScope,
+  #[string] module_specifier: String,
+) -> Result<v8::Global<v8::Value>, Error> {
+  let module_map_rc = JsRealm::module_map_from(scope);
+
+  let lazy_esm = module_map_rc.data.borrow().lazy_esm.clone();
+  let loader = crate::modules::LazyEsmModuleLoader::new(lazy_esm);
+
+  let mod_ns = futures::executor::block_on(async {
+    let specifier = crate::ModuleSpecifier::parse(&module_specifier)?;
```
Comment (bartlomieju) [inline core/ops_builtin_v8.rs:50]: Factor out most of this function to a helper method on the `ModuleMap`. Also currently it's not idempotent and I think it will crash it once tries to load the same module twice.

File: core/runtime/jsruntime.rs [hunk 14]
```diff
@@ -946,6 +946,9 @@ impl JsRuntime {
             file_source.load()?,
           )?;
         }
+
+        module_map
+          .add_lazy_loaded_esm_sources(extension.get_lazy_loaded_esm_sources());
```
Comment (bartlomieju) [inline core/runtime/jsruntime.rs:951]: Maybe move it to the top of the loop; add a comment.

**Review by bartlomieju — COMMENTED** [review]
Drive-by self-review...

**Comment by bartlomieju** [comment]
Pretty much stuck on how to handle primordials now...

**Comment by bartlomieju** [comment]
Opened https://github.com/denoland/deno_core/issues/363 to solve the primordials problem.

File: core/modules/loaders.rs [hunk 15]
```diff
@@ -171,6 +172,62 @@ impl ModuleLoader for ExtModuleLoader {
...
+  fn load(
+    &self,
+    specifier: &ModuleSpecifier,
+    _maybe_referrer: Option<&ModuleSpecifier>,
+    _is_dyn_import: bool,
+  ) -> Pin<Box<ModuleSourceFuture>> {
+    let sources = self.sources.borrow();
+    let source = match sources.get(specifier.as_str()) {
+      Some(source) => source,
+      None => return futures::future::err(anyhow!("Specifier \"{}\" cannot be lazy-loaded as it was not included in the snapshot.", specifier)).boxed_local(),
```
Comment (mmastrac) [inline core/modules/loaders.rs:209]: Are they actually snapshotted, or are they just made available?

File: core/modules/tests.rs [hunk 16]
```diff
@@ -410,6 +410,30 @@ fn test_mods() {
...
+    extensions: vec![test_ext::init_ops_and_esm()],
+    ..Default::default()
+  });
+
+  runtime
+    .execute_script_static(
+      "setup.js",
+      r#"
+      Deno.core.print("1\n");
+      const module = Deno.core.ops.op_lazy_load_esm("ext:test_ext/lazy_loaded.js");
```
Comment (mmastrac) [inline core/modules/tests.rs:427]: Can we add this as built-in on `core`? Maybe `Deno.core.importSync`

File: core/modules/loaders.rs (hunk 15 above)
Comment (bartlomieju) [reply core/modules/loaders.rs:209]: Ooops, seems like I didn't change the error message completely. It was meant to say `it was not inlucded in the binary`.

**Review by mmastrac — APPROVED** [review]
